// ngx-toasta/esm2020/lib/toasta.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ToastaComponent } from './toasta.component';
import { ToastComponent } from './toast.component';
import { SafeHtmlPipe } from './shared';
import { ToastaService, ToastaConfig, toastaServiceFactory } from './toasta.service';
import * as i0 from "@angular/core";
export let providers = [
    ToastaConfig,
    { provide: ToastaService, useFactory: toastaServiceFactory, deps: [ToastaConfig] }
];
export class ToastaModule {
    static forRoot() {
        return {
            ngModule: ToastaModule,
            providers
        };
    }
}
ToastaModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.2", ngImport: i0, type: ToastaModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
ToastaModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.2", ngImport: i0, type: ToastaModule, declarations: [ToastComponent, ToastaComponent, SafeHtmlPipe], imports: [CommonModule], exports: [ToastComponent, ToastaComponent] });
ToastaModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.2", ngImport: i0, type: ToastaModule, providers: providers, imports: [CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.2", ngImport: i0, type: ToastaModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [ToastComponent, ToastaComponent, SafeHtmlPipe],
                    exports: [ToastComponent, ToastaComponent],
                    providers
                }]
        }] });
